### ISSUE TITLE
[QA] UI maps credentials to a malformed URL, when protocol is specified.

### ISSUE THREAD

**Issue by jnweiger** [owncloud/search_elastic#258]
New scope: HTTPS support is needed for Elastic.

Seen in search_elastic 2.1.0 RC1

* admin enables app search_elastic
* admin goes to the 'search' settings and enters host and port.
* `occ config:app:get search_elastic servers`
```
http://elastic:ua9116rq7e@172.17.0.3:9200
```
* Switch the Authentiction  dropdown from 'Select Authentication' to 'User and password'
* enter correct username and password. Click save settings, the indicator remains red and the message says cannot connect.
![image](https://user-images.githubusercontent.com/1108546/155594491-df3f50f7-dd93-49e2-8b0f-7a52b026860e.png)

* The server URL is now malformed: `occ config:app:get search_elastic servers`
```
elastic:ua9116rq7e@http://172.17.0.3:9200
```

#### Expected behaviour
* Credentials follow after `://`

#### Attempted Workaround
* embed the credentials in the URL and do not select 'Username and Password'

![image](https://user-images.githubusercontent.com/1108546/155595061-8b191c71-a82b-48a1-8cbe-a210f18a4013.png)
----

This stores the URL correctly in the database (`http://elastic:ua9116rq7e@172.17.0.3:9200`), but after a refresh the web page displays garbled fields:

![image](https://user-images.githubusercontent.com/1108546/155595210-9b35662e-0e85-4435-8e40-2e110c5c0f1b.png)
----

Still an index cannot be created:
![image](https://user-images.githubusercontent.com/1108546/155595470-c20da48d-0ac1-44a5-acdd-38c39e4e8ceb.png)

#### Actual workaround
* must not specify a protocol
  - (we are not prepared to handle HTTPS then?)
---
![image](https://user-images.githubusercontent.com/1108546/155595865-a2411ea7-75ea-4d96-9a48-935d18d116d0.png)

**Comment by hodyroff** [comment]
not a regression, needs checking later

**Comment by hodyroff** [comment]
Seems there is no workaround and this is mandatory for the authentication scheme. Blocker.

**Comment by xoxys** [comment]
@C0rby had some objections as well... IIRC reading back the password to the UI (even if it is hidden) isn't recommended https://github.com/owncloud/search_elastic/blob/master/js/settings/admin.js#L12-L19

So please coordinate with @C0rby if required...

**Comment by xoxys** [comment]
Summary: The URL parameter needs a proper handling for URL schemas (https/http/none)

**Comment by ahherrera** [comment]
@jnweiger I have been working on this ticket and would like to know if the protocol could be a select with two options (HTTP, HTTPS)?
@C0rby I don't understand the behavior you want to have with the password field. Could you explain me a little more in detail how the password field should behave?
Regards!!

**Comment by xoxys** [comment]
The protocol should IMO be part of the url and not a separate field. Php should provide a function to properly parse/split and build urls instead of the manual split and concatenation.

**Comment by C0rby** [comment]
> @jnweiger I have been working on this ticket and would like to know if the protocol could be a select with two options (HTTP, HTTPS)? @C0rby I don't understand the behavior you want to have with the password field. Could you explain me a little more in detail how the password field should behave? Regards!!

Well the least that shoud be done is that the password field is URL encoded before sending the request.
Imagine if someone has '#' in their password, this would completely break the url when we send the credentials in the URL. Not so much a problem when sending the credentials properly via headers.

But generally the URL handling needs to be made more resilient. 
In a normal case the admin sets for example the url: 10.10.10.161:9200
The request will be sent to user:password@10.10.10.161:9200/oc-randomchars
But if I now instead enter the url: 10.10.10.161:9200/somepath
The request will be sent to user:password@10.10.10.161:9200/somepath/oc-randomchars
Everything good so far...
But now I enter the url: 10.10.10.161:9200/somepath#
The request will be sent to user:password@10.10.10.161:9200/somepath (without our appended path /oc-randomchars)

That is why I say that we need to parse the URL and then correctly append the path.

**Comment by ahherrera** [comment]
> > @jnweiger I have been working on this ticket and would like to know if the protocol could be a select with two options (HTTP, HTTPS)? @C0rby I don't understand the behavior you want to have with the password field. Could you explain me a little more in detail how the password field should behave? Regards!!
> 
> Well the least that shoud be done is that the password field is URL encoded before sending the request. Imagine if someone has '#' in their password, this would completely break the url when we send the credentials in the URL. Not so much a problem when sending the credentials properly via headers.
> 
> But generally the URL handling needs to be made more resilient. In a normal case the admin sets for example the url: 10.10.10.161:9200 The request will be sent to user:password@10.10.10.161:9200/oc-randomchars But if I now instead enter the url: 10.10.10.161:9200/somepath The request will be sent to user:password@10.10.10.161:9200/somepath/oc-randomchars Everything good so far... But now I enter the url: 10.10.10.161:9200/somepath# The request will be sent to user:password@10.10.10.161:9200/somepath (without our appended path /oc-randomchars)
> 
> That is why I say that we need to parse the URL and then correctly append the path.

@C0rby I think that after the connection port should not go anything else because in that field we are configuring the connection to the server, not requests to the elastic. Do you think there is any scenario where you have parts of the url after the port?

**Comment by C0rby** [comment]
It's not about what you intend a user to put there but what a (malicious) user COULD put there. This could be abused to scan the internal network for hosts/applications. Even if it would take a while with that method.

**Comment by C0rby** [comment]
And also yes, I think there could be setups where the elastic server is reachable at a subpath like 10.10.10.161/elastic.

**Comment by jnweiger** [comment]
Confirmed fixed in 2.2.0-rc1
